src/mq.h

```diff
@@ -0,0 +1,121 @@
+
+#ifndef MQ_H_
+#define MQ_H_
+
+#include <assert.h>
+#include <mutex>
+#include "deno.h"
+
+namespace deno {
+
+class MessageQueue {
+  struct Message {
+    deno_buf buf = DENO_BUF_INIT;
+    Message* next;
+  };
+
+  typedef std::mutex Mutex;
+
+  Message* head_;
+  Message* tail_;
+
+  Mutex mutex_;
+  std::condition_variable cv_;
+  bool reader_is_blocked_;
+
+ public:
+  MessageQueue()
+      : head_(nullptr),
+        tail_(nullptr),
+        mutex_(),
+        cv_(),
+        reader_is_blocked_(false) {}
+
+  void Send(deno_buf* buf, bool nowake = false) {
+    auto m = new Message;
+    deno_buf_move_into(&m->buf, buf);
+
+    std::unique_lock<Mutex> lock(mutex_);
+
+    if (head_ == nullptr) {
```

Comment:
Maybe comment that this is a standard link list append. Or maybe use `std::list<deno_buf>`?